chain/actors/actor_storagemarket.go

```diff
@@ -0,0 +1,574 @@
+		vmctx.Message().From: balance,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	self.Balances = bcid
+
+	nroot, err := vmctx.Storage().Put(&self)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, vmctx.Storage().Commit(old, nroot)
+}
+
+func (sma StorageMarketActor) AddBalance(act *types.Actor, vmctx types.VMContext, params *struct{}) ([]byte, ActorError) {
+	var self StorageMarketState
+	old := vmctx.Storage().GetHead()
+	if err := vmctx.Storage().Get(old, &self); err != nil {
+		return nil, err
+	}
+
+	b, bnd, err := GetMarketBalances(vmctx.Context(), vmctx.Ipld(), self.Balances, vmctx.Message().From)
+	if err != nil {
+		return nil, aerrors.Wrap(err, "could not get balance")
+	}
+
+	balance := b[0]
+
+	balance.Available = types.BigAdd(balance.Available, vmctx.Message().Value)
+
+	bcid, err := setMarketBalances(vmctx, bnd, map[address.Address]StorageParticipantBalance{
+		vmctx.Message().From: balance,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	self.Balances = bcid
+
+	nroot, err := vmctx.Storage().Put(&self)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, vmctx.Storage().Commit(old, nroot)
+}
+
+func setMarketBalances(vmctx types.VMContext, nd *hamt.Node, set map[address.Address]StorageParticipantBalance) (cid.Cid, ActorError) {
+	for addr, b := range set {
+		if err := nd.Set(vmctx.Context(), string(addr.Bytes()), b); err != nil {
+			return cid.Undef, aerrors.HandleExternalError(err, "setting new balance")
+		}
+	}
+	if err := nd.Flush(vmctx.Context()); err != nil {
+		return cid.Undef, aerrors.HandleExternalError(err, "flushing balance hamt")
+	}
+
+	c, err := vmctx.Ipld().Put(vmctx.Context(), nd)
+	if err != nil {
+		return cid.Undef, aerrors.HandleExternalError(err, "failed to balances storage")
+	}
+	return c, nil
+}
+
+func GetMarketBalances(ctx context.Context, store *hamt.CborIpldStore, rcid cid.Cid, addrs ...address.Address) ([]StorageParticipantBalance, *hamt.Node, ActorError) {
+	nd, err := hamt.LoadNode(ctx, store, rcid)
+	if err != nil {
+		return nil, nil, aerrors.HandleExternalError(err, "failed to load miner set")
+	}
+
+	out := make([]StorageParticipantBalance, len(addrs))
+
+	for i, a := range addrs {
+		var balance StorageParticipantBalance
+		err = nd.Find(ctx, string(a.Bytes()), &balance)
+		switch err {
+		case hamt.ErrNotFound:
+			out[i] = StorageParticipantBalance{
+				Locked:    types.NewInt(0),
+				Available: types.NewInt(0),
+			}
+		case nil:
+			out[i] = balance
+		default:
+			return nil, nil, aerrors.HandleExternalError(err, "failed to do set lookup")
+		}
+
+	}
+
+	return out, nd, nil
+}
+
+/*
+func (sma StorageMarketActor) CheckLockedBalance(act *types.Actor, vmctx types.VMContext, params *struct{}) ([]byte, ActorError) {
+
+}
+*/
+
+type PublishStorageDealsParams struct {
+	Deals []StorageDeal
+}
+
+type PublishStorageDealResponse struct {
+	DealIDs []uint64
+}
+
+// TODO: spec says 'call by CommitSector in StorageMiningSubsystem', and then
```

Comment:
will fix this, this should be called before `CommitSector`. Calling PublishStorageDeals from CommitSector is an optimization that can be done later.

Reply:
This should return `DealID` for each `StorageDeal` to be stored in `StorageMarketActor` and in `OnChainSealVerifyInfo`